include/nbl/builtin/hlsl/math/integral/gauss_legendre.hlsl

```diff
@@ -0,0 +1,427 @@
+// Copyright (C) 2018-2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_MATH_INTEGRAL_GAUSS_LEGENDRE_INCLUDED_
+#define _NBL_BUILTIN_HLSL_MATH_INTEGRAL_GAUSS_LEGENDRE_INCLUDED_
+
+// TODO FIX:
+// #include <nbl/builtin/hlsl/cpp_compat/cpp_compat.hlsl>
+#define NBL_CONSTEXPR const static
+#define NBL_CONSTEXPR_STATIC_INLINE const static
+
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace math
+{
+    namespace integral
+    {
+        template<int Order, typename float_t>
+        struct GaussLegendreValues 
+        {};
+
+        template<int Order, typename float_t, class IntegrandFunc>
+        struct GaussLegendreIntegration
+        {
+            static float_t calculateIntegral(IntegrandFunc func, float_t start, float_t end)
```

Comment:
just "in case" your function is "stateful" you want to be using `NBL_CONST_REF_ARG(IntegrandFunc)` so that HLSL is using `inout` and C++ is using `const&`

Reply:
using `NBL_CONST_REF_ARG` is blocked by a bug fix from Arek and Atil. I'll add a TODO